[linustws] iP

linustws opened the pull request:
# Duke (BorzAI Edition)
> "It's not the load that breaks you down, it's the way you carry it." - Lou Holtz [(source)](https://quotes4sharing.com/3367/)

Let BorzAI unleash the full potential of your mind by memorising things you need to do so you don't have to. It's,
- simple
- fast to learn
- ~~EASY~~ _SUPER EASY_ to use

All you have to do is,
1. download it from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ&ab_channel=RickAstley).
2. install it.
3. run it.
4. let it handle your tasks for you :star_struck:

And it is **FREE!**

Features:

 - [x] Adding tasks
 - [x] Removing tasks
 - [x] Show list
 - [x] Marking and unmarking tasks
 - [x] Filter by date
 - [x] Filter by keyword
 - [ ] Clearing all tasks (coming soon)   
 - [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Duke {
    public static void main(String[] args) {
        new Duke("./data/tasks.ser").run();
    }
}
```

Commit:
Add Gradle support

linustws committed:
Greet, Echo, Exit

linustws committed:
Add, List

linustws committed:
Mark as Done

linustws committed:
ToDos, Events, Deadlines

linustws committed:
A-TextUiTesting

linustws committed:
A-TextUiTesting

linustws committed:
Level-5

linustws committed:
Level-6

linustws committed:
A-Enums

linustws committed:
Level-7

linustws committed:
Introduce save feature

linustws committed:
Introduce date and time

linustws committed:
Merge tag 'Level-7'

Level-7

linustws committed:
Merge tag 'Level-8'

linustws committed:
A-MoreOOP

linustws committed:
Divide classes into packages

linustws committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

linustws committed:
Set up gradle

linustws committed:
Add JUnit Tests

linustws committed:
Add checkstyle to gradle

linustws committed:
Add JavaDoc comments

linustws committed:
Tweak code to comply with coding standard

linustws committed:
Add filter tasks by keyword

linustws committed:
Merge branch 'branch-A-Javadoc'

linustws committed:
Merge branch 'branch-A-CodingStandard'

linustws committed:
Merge branch 'branch-Level-9'

linustws committed:
Use CheckStyle

anchengyang commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,125 @@
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
+
```
Maybe this empty line could be removed

anchengyang commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,125 @@
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] parts = fullCommand.split(" ", 2);
+        int indexInput;
+        switch(parts[0].trim()) {
```
switch cases statements do not need an indent as per the [coding convention](https://se-education.org/guides/conventions/java/intermediate.html#layout)

anchengyang reviewed (commented):
Just a few coding convention issues  but other than that good job on the OOP design and the commenting.😄

anchengyang commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,152 @@
+        LocalDate endDate = null;
+        LocalDateTime endDateTime = null;
+        DateTimeFormatter[] formatters = {
+                DateTimeFormatter.ofPattern("ddMMyyyy HHmm"),
```
Good job on including so many datetime formats for Event class even when it is not one of the required increments 👍

anchengyang commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,142 @@
+package duke.ui;
```
nice UI class!

anchengyang commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,120 @@
+package duke;
```
very comprehensive test cases that tests for almost all features

anchengyang commented on src/test/java/duke/ui/UiTest.java:
```diff
@@ -0,0 +1,65 @@
+package duke.ui;
+
+import org.junit.jupiter.api.Test;
+
```
Maybe all these empty lines between import statements can be removed?

anchengyang reviewed (commented):
Some positive comments that I found quite helpful for myself!

linustws commented:
> Just a few coding convention issues but other than that good job on the OOP design and the commenting.😄

Thanks for the feedback!!

linustws committed:
Add GUI to Duke

linustws committed:
Use Varargs

linustws committed:
Merge branch 'branch-A-CheckStyle'

linustws committed:
Merge branch 'branch-Level-10'

linustws committed:
Merge branch 'branch-A-Varargs'

linustws committed:
Use Assertions

No assertions used.

Assertions can be used to enforce assumptions that are critical to the
correctness of the code.

Let's,
* add assertions to check that the index the user entered for mark,
  unmark and delete commands is 1 or greater.
* insert the assertion before tasks.get(index) to show the assertion
  error before the DukeException is thrown.
* must enable assertions in the run configuration to display assertion
  errors, but this will prevent displaying the DukeException error
  message.

linustws committed:
Improve code quality: extract formatting of dates

Deadline and Events take in strings as arguments and have to convert
them to valid dates within their constructors.

Deadlines and Events should take in valid dates in the first place
without the conversion of strings.

Let's,
* extract the formatting of dates to be done in the parser class, where
  the DukeExceptions will be thrown if the date entered by user is
  invalid.
* create methods to convert string dates to LocalDateTime and LocalDate
  to reduce code duplication.

This way, DateTimeFormatters can be in one place without being in many
classes. Code will also be more readable and maintainable.

linustws committed:
Use Streams

linustws committed:
Merge pull request #1 from linustws/branch-A-Assertions

Use Assertions

linustws committed:
Merge branch 'branch-A-CodeQuality'

linustws committed:
Merge branch 'branch-A-Streams'

linustws committed:
Add GitHub Actions

linustws committed:
Merge pull request #4 from linustws/linustws-patch-1

Add GitHub Actions

linustws committed:
Revert "Add GitHub Actions"

linustws committed:
Merge pull request #5 from linustws/revert-4-linustws-patch-1

Revert "Add GitHub Actions"

linustws committed:
Sort tasks by description/date/type/status

linustws committed:
Merge pull request #6 from linustws/branch-C-Sort

Sort tasks by description/date/type/status

linustws committed:
Improve the GUI

linustws committed:
Merge branch 'branch-A-BetterGui'

linustws committed:
Give a unique personality

linustws committed:
Merge branch 'branch-A-Personality'

linustws committed:
Handle more errors

linustws committed:
Merge branch 'branch-A-MoreErrorHandling'

linustws committed:
Write more automated tests

linustws committed:
Merge branch 'branch-A-MoreTesting'

linustws committed:
Clean up

linustws committed:
Clean up

linustws committed:
Clean up

linustws committed:
Clean up

linustws committed:
Add a User Guide

linustws committed:
Merge branch 'branch-A-UserGuide'

linustws committed:
Create _config.yml and add theme

linustws committed:
Merge pull request #8 from linustws/linustws-patch-2

Create _config.yml and add theme

linustws committed:
Clean up

linustws committed:
Clean up

linustws committed:
Change Gradle build

linustws committed:
Update user guide

linustws committed:
Checkstyle

linustws committed:
Update README.md

linustws committed:
Update GitHub Pages

linustws committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

linustws committed:
Change UG and README

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Change UG

linustws committed:
Update README.md
